### PR TITLE
[ISSUE #453] OpenSchema integration with EventMesh

### PR DESCRIPTION
**Pull request by yzhao244**
Fixes to child ISSUE#453 of its parent ISSUE#339.

### Motivation

- Support Schema Registry feature in EventMesh by integrating with OpenSchema Specifications 

*Explain why you want to make the changes and what problem you're trying to solve.*

- The purpose of this pull request is for introducing Schema Registry as part of the EventMesh.
The Schema Registry is a central repository with RESTful interfaces for developers to define and register standard schemas. Addresses the problem of inconsistent or incompatible data(event) format between event producer and event consumer.

### Modifications

- Added eventmesh-store-api module which is an abstract layer abstracts the CRUD capability of the schema registry into this module.
- Added eventmesh-store-h2 module which contains the actual implementation
- Modified eventmesh-runtime which exposes three new subject APIs (schemaregistry/createSubject, /schemaregistry/readSubjectByName and /schemaregistry/showAllSubjectNames) via ClientManageController.

### Documentation

- Does this pull request introduce a new feature? yes
- If yes, how is the feature documented? Another follow up issue or PR will be created for adding the documentation
- If a feature is not applicable for documentation, explain why?
- If a feature is not documented yet in this PR, please create a followup issue for adding the documentation

### CODE REVIEW

**Comment by yzhao244** [comment]
- POST http://localhost:10106/schemaregistry/createSubject
The following is the payload of creating a new subject 
```
{
	"subject": "test-topic",
	"namespace": "org.apache.rocketmq",
	"tenant": "messaging/rocketmq",
	"app": "rocketmq",
	"description": "rocketmq user information",
	"compatibility": "NONE",
	"status": "deprecated"
}
```
- GET http://localhost:10106/schemaregistry/readSubjectByName?subject=test-topic
- GET http://localhost:10106/schemaregistry/showAllSubjectNames

**Comment by qqeasonchen** [comment]
@yzhao244 License check seems failed.

**Comment by ruanwenjun** [comment]
@yzhao244 It might be better to add a new module such like `eventmesh-store-plugin`, and make the `eventmesh-store-h2` as a submodule of `eventmesh-store-plugin`, this makes it easier to manage plugins.

And we have recently enhanced the SPI module #419, you can use `EventMeshExtensionFactory.getExtension(Class<T> extensionType, String extensionName)` to get plugin instance rather than using `ServiceLoader`. You can find some detail of SPI in `eventmesh-connector-rocketmq` module.

**Comment by qqeasonchen** [comment]
> @yzhao244 It might be better to add a new module such like `eventmesh-store-plugin`, and make the `eventmesh-store-h2` as a submodule of `eventmesh-store-plugin`, this makes it easier to manage plugins.
> 
> And we have recently enhanced the SPI module #419, you can use `EventMeshExtensionFactory.getExtension(Class<T> extensionType, String extensionName)` to get plugin instance rather than using `ServiceLoader`. You can find some detail of SPI in `eventmesh-connector-rocketmq` module.

yes, @ruanwenjun It seems also needs a eventmesh-connector-plugin moudle.

File: eventmesh-store-h2/src/main/java/org/apache/eventmesh/store/h2/schema/util/DBDataSource.java [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+import java.sql.SQLException;
+
+public class DBDataSource {
+    private static BasicDataSource ds = null;
+
+    private DBDataSource() {
+
+    }
+
+    public static synchronized DBDataSource createDataSource(DBConfiguration dbConfig) {
```
Comment (ruanwenjun) [inline eventmesh-store-h2/src/main/java/org/apache/eventmesh/store/h2/schema/util/DBDataSource.java:16]: The `DBDataSource` is singleton?

**Comment by ruanwenjun** [comment]
@qqeasonchen You are right, the `eventmesh-connector-plugin` is at this pr #433

**Comment by yzhao244** [comment]
@qqeasonchen @ruanwenjun Thanks a lot for the review comments. I will start to address them :)

File: eventmesh-runtime/src/main/java/org/apache/eventmesh/runtime/admin/handler/CreateSubjectHandler.java [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+
+import com.fasterxml.jackson.databind.DeserializationFeature;
+import com.fasterxml.jackson.databind.SerializationFeature;
+import com.fasterxml.jackson.annotation.JsonInclude;
+
+public class CreateSubjectHandler implements HttpHandler {
+	
+	private Logger logger = LoggerFactory.getLogger(CreateSubjectHandler.class);
+	
+	private static ObjectMapper jsonMapper;
```
[the remark this answers is not in the document]
Comment (yzhao244) [reply eventmesh-runtime/src/main/java/org/apache/eventmesh/runtime/admin/handler/CreateSubjectHandler.java:44]: @ruanwenjun just to confirm :) .. do you mean by creating a new utility class called "JsonUtils" under path (org.apache.eventmesh.runtime.util) which leverages ObjectMapper for JSON serialization or deserialization?

[the remark this answers is not in the document]
Comment (ruanwenjun) [reply eventmesh-runtime/src/main/java/org/apache/eventmesh/runtime/admin/handler/CreateSubjectHandler.java:44]: Maybe it can be placed at common module

**Comment by Jackzeng1224** [comment]
Hello, @yzhao244 !  I looked  interface at the development, it seems that some interfaces are missing. I refer to https://github.com/openmessaging/openschema/blob/master/spec.md
for instance: 
    7.1.2 
    7.2.4 
    7.2.6 
    7.2.7

**Comment by yzhao244** [comment]
> Hello, @yzhao244 ! I looked interface at the development, it seems that some interfaces are missing. I refer to https://github.com/openmessaging/openschema/blob/master/spec.md
> for instance:
> 7.1.2
> 7.2.4
> 7.2.6
> 7.2.7

@Jackzeng1224 Thanks for the question.  Yes, you are right. The current commit so far only contains three Subject related APIs such as create subject, readySubjectbyname and showAllSubjects which have been mentioned in my first comment of this PR. :) . However, I am going to implement the rest of missing interfaces as soon as we are okay with the new modules(eventmesh-store-api and eventmesh-store-plugin), because I can follow new modules standards to code/implement the rest of interfaces. lol

File: eventmesh-store-h2/src/main/java/org/apache/eventmesh/store/h2/schema/util/DBDataSource.java (hunk 1 above)
Comment (yzhao244) [reply eventmesh-store-h2/src/main/java/org/apache/eventmesh/store/h2/schema/util/DBDataSource.java]: Yes, DBDataSource is singleton. Basically, it performs init, start and shutdown as part of EventMeshServer init, start and shutdown process.

File: eventmesh-store-api/src/main/java/org/apache/eventmesh/store/api/openschema/request/ConfigUpdateRequest.java [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+@JsonInclude(JsonInclude.Include.NON_EMPTY)
+@JsonIgnoreProperties(ignoreUnknown = true)
+public class ConfigUpdateRequest {
+
+	  private String compatibilityLevel;
+
+	 /* @ApiModelProperty(value = "Compatability Level",
+	      allowableValues = "BACKWARD, BACKWARD_TRANSITIVE, FORWARD, FORWARD_TRANSITIVE, FULL, "
+	          + "FULL_TRANSITIVE, NONE")*/
+	  @JsonProperty("compatibility")
```
[the remark this answers is not in the document]
Comment (yzhao244) [reply eventmesh-store-api/src/main/java/org/apache/eventmesh/store/api/openschema/request/ConfigUpdateRequest.java:30]: fixed

File: eventmesh-store-api/src/main/java/org/apache/eventmesh/store/api/openschema/request/ConfigUpdateRequest.java [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+import com.fasterxml.jackson.annotation.JsonInclude;
+import com.fasterxml.jackson.annotation.JsonProperty;
+
+@JsonInclude(JsonInclude.Include.NON_EMPTY)
+@JsonIgnoreProperties(ignoreUnknown = true)
+public class ConfigUpdateRequest {
+
+	  private String compatibilityLevel;
+
+	 /* @ApiModelProperty(value = "Compatability Level",
```
[the remark this answers is not in the document]
Comment (yzhao244) [reply eventmesh-store-api/src/main/java/org/apache/eventmesh/store/api/openschema/request/ConfigUpdateRequest.java]: fixed

File: eventmesh-runtime/src/main/java/org/apache/eventmesh/runtime/admin/handler/CreateSubjectHandler.java [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+                try {
+                    out.close();
+                } catch (IOException e) {
+                    logger.warn("out close failed...", e);
+                }
+            }
+        }
+    }
+    
+    private SchemaService getSchemaService() {
```
[the remark this answers is not in the document]
Comment (yzhao244) [reply eventmesh-runtime/src/main/java/org/apache/eventmesh/runtime/admin/handler/CreateSubjectHandler.java]: fixed

File: eventmesh-runtime/src/main/java/org/apache/eventmesh/runtime/admin/handler/CreateSubjectHandler.java [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+                return;
+            }
+        } catch (ServiceException e) {            	
+        	httpExchange.getResponseHeaders().add("Content-Type", "appication/json");
+            httpExchange.sendResponseHeaders(500, 0);                            
+            result = jsonMapper.writeValueAsString(e.getErrorResponse());
+            logger.error(result);
+            out.write(result.getBytes());
+            return;
+        } finally {
```
[the remark this answers is not in the document]
Comment (yzhao244) [reply eventmesh-runtime/src/main/java/org/apache/eventmesh/runtime/admin/handler/CreateSubjectHandler.java:91]: In catch block, out.write(result.getBytes()); is used to return error response.. Therefore, out.write(result.getBytes()) becomes undefined if defining "OutputStream out" using Try with resource

File: eventmesh-runtime/src/main/java/org/apache/eventmesh/runtime/admin/handler/CreateSubjectHandler.java [hunk 7]
```diff
@@ -0,0 +1,94 @@
+package org.apache.eventmesh.runtime.admin.handler;
```
[the remark this answers is not in the document]
Comment (yzhao244) [reply eventmesh-runtime/src/main/java/org/apache/eventmesh/runtime/admin/handler/CreateSubjectHandler.java:18]: fixed license header

File: eventmesh-runtime/src/main/java/org/apache/eventmesh/runtime/admin/controller/ClientManageController.java [hunk 8]
```diff
@@ -57,6 +60,10 @@ public void start() throws IOException {
         server.createContext("/clientManage/redirectClientByPath", new RedirectClientByPathHandler(eventMeshTCPServer));
         server.createContext("/clientManage/redirectClientByIpPort", new RedirectClientByIpPortHandler(eventMeshTCPServer));
         server.createContext("/clientManage/showListenClientByTopic", new ShowListenClientByTopicHandler(eventMeshTCPServer));
+        
+        server.createContext("/schemaregistry/createSubject", new CreateSubjectHandler());
```
[the remark this answers is not in the document]
Comment (yzhao244) [reply eventmesh-runtime/src/main/java/org/apache/eventmesh/runtime/admin/controller/ClientManageController.java]: fixed

File: eventmesh-runtime/build.gradle [hunk 9]
```diff
@@ -29,8 +29,12 @@ List open_message = [
         "io.openmessaging:openmessaging-api:2.2.1-pubsub"
 ]
 
+List h2_database = [
```
[the remark this answers is not in the document]
Comment (yzhao244) [reply eventmesh-runtime/build.gradle]: fixed

**Comment by yzhao244** [comment]
> @yzhao244 It might be better to add a new module such like `eventmesh-store-plugin`, and make the `eventmesh-store-h2` as a submodule of `eventmesh-store-plugin`, this makes it easier to manage plugins.
> 
> And we have recently enhanced the SPI module #419, you can use `EventMeshExtensionFactory.getExtension(Class<T> extensionType, String extensionName)` to get plugin instance rather than using `ServiceLoader`. You can find some detail of SPI in `eventmesh-connector-rocketmq` module.

@ruanwenjun Thanks for the suggestion. I have made eventmesh-store-h2 as a submodule of eventmesh-store-plugin module  and enhanced my implementation by using EventMeshExtensionFactory.getExtension instead of ServiceLoader. :)

File: eventmesh-store-h2/src/main/java/org/apache/eventmesh/store/h2/schema/util/DBDataSource.java (hunk 1 above)
Comment (ruanwenjun) [reply eventmesh-store-h2/src/main/java/org/apache/eventmesh/store/h2/schema/util/DBDataSource.java]: But when we execute `createDataSource` method twice, we will get two `DBDataSource` instances, this is confusion, I think you would better make this class to be a real singleton.

**Comment by Jackzeng1224** [comment]
> > Hello, @yzhao244 ! I looked interface at the development, it seems that some interfaces are missing. I refer to https://github.com/openmessaging/openschema/blob/master/spec.md
> > for instance:
> > 7.1.2
> > 7.2.4
> > 7.2.6
> > 7.2.7
> 
> @Jackzeng1224 Thanks for the question. Yes, you are right. The current commit so far only contains three Subject related APIs such as create subject, readySubjectbyname and showAllSubjects which have been mentioned in my first comment of this PR. :) . However, I am going to implement the rest of missing interfaces as soon as we are okay with the new modules(eventmesh-store-api and eventmesh-store-plugin), because I can follow new modules standards to code/implement the rest of interfaces. lol

Thanks for answering questions! Looking forward to eventmesh-store-plugin module

**Comment by Jackzeng1224** [comment]
@yzhao244  hello，I found a problem with createDataSource. First, the new H2SchemaAdapter is initialized.
image
DBDataSource.createDataSource(dbConfig) in the start() method.
There should be a problem with the createDataSource() method. For example: ds, new DBDataSource(). It is different from other methods of creating data sources.
image

File: eventmesh-store-api/gradle.properties [hunk 10]
```diff
@@ -0,0 +1,21 @@
...
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+group=org.apache.eventmesh
+version=1.2.0-SNAPSHOT
+jdk=1.7
```
[the remark this answers is not in the document]
Comment (yzhao244) [reply eventmesh-store-api/gradle.properties]: fixed

File: eventmesh-store-api/src/main/java/org/apache/eventmesh/store/api/openschema/common/CompatibilityLevel.java [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+public enum CompatibilityLevel {
+	  NONE,
+	  BACKWARD,
+	  BACKWARD_TRANSITIVE,
+	  FORWARD,
+	  FORWARD_TRANSITIVE,
+	  FULL,
+	  FULL_TRANSITIVE;
+
+	  public final String name;
```
[the remark this answers is not in the document]
Comment (yzhao244) [reply eventmesh-store-api/src/main/java/org/apache/eventmesh/store/api/openschema/common/CompatibilityLevel.java]: fixed

File: eventmesh-store-api/src/main/java/org/apache/eventmesh/store/api/openschema/common/ServiceException.java [hunk 12]
```diff
@@ -0,0 +1,50 @@
...
+	    this.errorMessage = serviceError.getMessage();
+	  }
+
+	  @JsonProperty("error_code")
+	  public int getErrorCode() {
+	    return errorCode;
+	  }
+
+	  @JsonProperty("error_code")
+	  public void setErrorCode(int error_code) {
```
[the remark this answers is not in the document]
Comment (yzhao244) [reply eventmesh-store-api/src/main/java/org/apache/eventmesh/store/api/openschema/common/ServiceException.java]: fixed

File: eventmesh-store-h2/src/main/java/org/apache/eventmesh/store/h2/schema/util/DBDataSource.java (hunk 1 above)
Comment (yzhao244) [reply eventmesh-store-h2/src/main/java/org/apache/eventmesh/store/h2/schema/util/DBDataSource.java]: You are right. I pushed fixes to DBDataSource class. thx

**Comment by Jackzeng1224** [comment]
@yzhao244  hello,I looked at DBDataSource(), this "ds" does not belong to the object. Did you look right?
image

**Comment by yzhao244** [comment]
> @yzhao244 hello,I looked at DBDataSource(), this "ds" does not belong to the object. Did you look right?
> image

You are right, I removed "static" when defining BasicDataSource ds.

**Comment by Jackzeng1224** [comment]
> > @yzhao244你好，我看了DBDataSource()，这个“ds”不属于对象。你看对了吗？
> > 图像
> 
> 你是对的，我在定义 BasicDataSource ds 时删除了“静态”。
Thanks! I will try again

**Comment by yzhao244** [comment]
@qqeasonchen @ruanwenjun Hi guys, do you have any concerns before merging this PR? Please let me know if I missed anything here. :)

**Comment by ruanwenjun** [comment]
> @qqeasonchen @ruanwenjun Hi guys, do you have any concerns before merging this PR? Please let me know if I missed anything here. :)

@yzhao244 Hi, I have a question, the h2 is a memory database, it seems doesn't support distributed, how can the different eventmesh-runtime sync the schema change?

**Comment by qqeasonchen** [comment]
> > @qqeasonchen @ruanwenjun Hi guys, do you have any concerns before merging this PR? Please let me know if I missed anything here. :)
> 
> @yzhao244 Hi, I have a question, the h2 is a memory database, it seems doesn't support distributed, how can the different eventmesh-runtime sync the schema change?



> @qqeasonchen @ruanwenjun Hi guys, do you have any concerns before merging this PR? Please let me know if I missed anything here. :)

1. Eventmesh-store-api can be moved into eventmesh-store-plugin, just like connector.
2. Did all the apis in the openSchema implement now?
3. Did the management apis implement now?
4. Did the sdk implement now?

**Comment by Jackzeng1224** [comment]
@yzhao244  hello,this problem occurs with the new receiving code.  "configurationWraper"  or "configurationWrapper"?
image

**Comment by yzhao244** [comment]
> > > @qqeasonchen @ruanwenjun Hi guys, do you have any concerns before merging this PR? Please let me know if I missed anything here. :)
> > 
> > 
> > @yzhao244 Hi, I have a question, the h2 is a memory database, it seems doesn't support distributed, how can the different eventmesh-runtime sync the schema change?
> 
> > @qqeasonchen @ruanwenjun Hi guys, do you have any concerns before merging this PR? Please let me know if I missed anything here. :)
> 

@qqeasonchen I answered your questions below. :)
> 1. Eventmesh-store-api can be moved into eventmesh-store-plugin, just like connector.
       Yes, I have pushed my latest changes which moved Eventmesh-store-api to be under eventmesh-store-plugin.
> 2. Did all the apis in the openSchema implement now?
        Not yet, I am thinking it is better delivering OpenSchema Integration in an incremental delivery fashion. :) .. In total, OpenSchema APIs can be seen as three groups.. /subject/ related APIs, /schema/ related APIs, /config/compatibility related APIs which I would suggest to deliver each group by individually separated PRs. The PR 434 currently delivers /subject/ related APIs.
       Three child issues are [Implement OpenSchema Subject APIs #453
](https://github.com/apache/incubator-eventmesh/issues/453), [Implement OpenSchema Schema APIs #454](https://github.com/apache/incubator-eventmesh/issues/454), [Implement OpenSchema Compatibility APIs #455](https://github.com/apache/incubator-eventmesh/issues/455)
> 3. Did the management apis implement now?
       Three new subject APIs (schemaregistry/createSubject, /schemaregistry/readSubjectByName and /schemaregistry/showAllSubjectNames) via ClientManageController. The rest of the OpenSchema APIs will be tracked/implemented by individually separated issues and PRs
> 4. Did the sdk implement now?
      Sorry, do you mean by eventmesh-sdk-java? All OpenSchema APIs are exposed as a RESTful API via webhook in ClientManageController.

**Comment by yzhao244** [comment]
> > @qqeasonchen @ruanwenjun Hi guys, do you have any concerns before merging this PR? Please let me know if I missed anything here. :)
> 
> @yzhao244 Hi, I have a question, the h2 is a memory database, it seems doesn't support distributed, how can the different eventmesh-runtime sync the schema change?

very good question.. Actually, h2 can be turned into a persistent database by using the following example configuration (datasource.url=jdbc:h2:file:~/subdirectory/demodb) which persists data in file system. 
Additionally,  database configuration/url can be easily changed to point to any MySQL database service as an example based on end-user's choice for different eventmesh-runtime instances to sync the schema change.  I am thinking to create separate issue for tracking documentation works for OpenSchema Integration and h2 plugin module. 
image

**Comment by qqeasonchen** [comment]
> > > > @qqeasonchen @ruanwenjun Hi guys, do you have any concerns before merging this PR? Please let me know if I missed anything here. :)
> > > 
> > > 
> > > @yzhao244 Hi, I have a question, the h2 is a memory database, it seems doesn't support distributed, how can the different eventmesh-runtime sync the schema change?
> > 
> > 
> > > @qqeasonchen @ruanwenjun Hi guys, do you have any concerns before merging this PR? Please let me know if I missed anything here. :)
> 
> @qqeasonchen I answered your questions below. :)
> 
> 1. Eventmesh-store-api can be moved into eventmesh-store-plugin, just like connector.
>    Yes, I have pushed my latest changes which moved Eventmesh-store-api to be under eventmesh-store-plugin.
> 2. Did all the apis in the openSchema implement now?
>    Not yet, I am thinking it is better delivering OpenSchema Integration in an incremental delivery fashion. :) .. In total, OpenSchema APIs can be seen as three groups.. /subject/ related APIs, /schema/ related APIs, /config/compatibility related APIs which I would suggest to deliver each group by individually separated PRs. The PR 434 currently delivers /subject/ related APIs.
>    Three child issues are [Implement OpenSchema Subject APIs #453
>    ](https://github.com/apache/incubator-eventmesh/issues/453), [Implement OpenSchema Schema APIs #454](https://github.com/apache/incubator-eventmesh/issues/454), [Implement OpenSchema Compatibility APIs #455](https://github.com/apache/incubator-eventmesh/issues/455)
> 3. Did the management apis implement now?
>    Three new subject APIs (schemaregistry/createSubject, /schemaregistry/readSubjectByName and /schemaregistry/showAllSubjectNames) via ClientManageController. The rest of the OpenSchema APIs will be tracked/implemented by individually separated issues and PRs
> 4. Did the sdk implement now?
>    Sorry, do you mean by eventmesh-sdk-java? All OpenSchema APIs are exposed as a RESTful API via webhook in ClientManageController.

@yzhao244 Individually separated PRs is good for me. But it is not very clear about how to use schema when pub/sub events, it seems eventmesh provides a management service for schema only in the request implementations, am i right?

**Comment by ruanwenjun** [comment]
@yzhao244 Great, you mean if we use `Mysql` then we don't need to consider Scheme synchronization, because all the runtimes connect to the same Mysql.  And if we use a public file system(e.g. NFS) to store h2 persistent file, then we also needn't consider Scheme synchronization, because the h2 help us to the job, am I right?
By the way, I don't know much about this, but I think your plan is ok. I think you would better send a mail to the community, let more people know your plan, and Apache community believes that all important decision need to be made in mail list.

**Comment by Jackzeng1224** [comment]
@yzhao244    hello，Can the module be tested and used normally now? or still to be modified.

**Comment by qqeasonchen** [comment]
@yzhao244 hi, I suggest split this pr into several small ones, so that it can easily be reviewed and merged, for example, eventmesh-rest、eventmesh-store-api、eventmesh-store-h2 ect.

**Comment by qqeasonchen** [comment]
@yzhao244 Could you please submmit this feature to branch openschema? i think eventmesh-rest can be merged to develop first.

**Comment by yzhao244** [comment]
@qqeasonchen @Jackzeng1224 @ruanwenjun Hi guys, I agree with @qqeasonchen that splitting this pr into smaller ones will be easier and safer for reviewing and merging each individual plugin/module. Therefore, I am thinking to close this PR and will submit separate PRs for each module after some modifications. 

Additionally, @ruanwenjun brought a good point regarding decision-making of important design. Should I create a design doc and push it to https://github.com/apache/incubator-eventmesh/tree/develop/docs/en/features (similar as eventmesh-stream-design.md which explains design in details and gets the whole community's awareness instead of only explaining designs under an issue) :)

**Comment by qqeasonchen** [comment]
@yzhao244 That would be good if the desing was clear.

**Pull request closed without merging**